firebase-firestore/src/main/java/com/google/firebase/firestore/model/value/ProtoValues.java

```diff
@@ -249,4 +250,92 @@ private static int compareMaps(MapValue left, MapValue right) {
     // Only equal if both iterators are exhausted.
     return Util.compareBooleans(iterator1.hasNext(), iterator2.hasNext());
   }
+
+  /** Generate the canonical ID for the provided field value (as used in Target serialization). */
+  public static String canonicalId(Value value) {
+    StringBuilder builder = new StringBuilder();
+    stringifyValue(builder, value);
+    return builder.toString();
+  }
+
+  // TODO(mrschmidt): Use in target serialization and migrate all existing TargetData
+  private static void stringifyValue(StringBuilder builder, Value value) {
+    switch (value.getValueTypeCase()) {
+      case NULL_VALUE:
+        builder.append("null");
+        break;
+      case BOOLEAN_VALUE:
+        builder.append(value.getBooleanValue());
+        break;
+      case INTEGER_VALUE:
+        builder.append(value.getIntegerValue());
+        break;
+      case DOUBLE_VALUE:
+        builder.append(value.getDoubleValue());
+        break;
+      case TIMESTAMP_VALUE:
+        stringifyTimestamp(builder, value.getTimestampValue());
+        break;
+      case STRING_VALUE:
+        builder.append(value.getStringValue());
+        break;
+      case BYTES_VALUE:
+        builder.append(Util.toDebugString(value.getBytesValue()));
+        break;
+      case REFERENCE_VALUE:
+        // TODO(mrschmidt): Use document key only
+        builder.append(value.getReferenceValue());
+        break;
+      case GEO_POINT_VALUE:
+        stringifyGeoPoint(builder, value.getGeoPointValue());
+        break;
+      case ARRAY_VALUE:
+        stringifyArray(builder, value.getArrayValue());
+        break;
+      case MAP_VALUE:
+        stringifyObject(builder, value.getMapValue());
+        break;
+      default:
+        throw fail("Invalid value type: " + value.getValueTypeCase());
+    }
+  }
+
+  private static void stringifyTimestamp(StringBuilder builder, Timestamp timestamp) {
+    builder.append(String.format("time(%s,%s)", timestamp.getSeconds(), timestamp.getNanos()));
+  }
+
+  private static void stringifyGeoPoint(StringBuilder builder, LatLng latLng) {
+    builder.append(String.format("geo(%s,%s)", latLng.getLatitude(), latLng.getLongitude()));
+  }
+
+  private static void stringifyObject(StringBuilder builder, MapValue mapValue) {
+    // Even though MapValue are likely sorted correctly based on their insertion order (e.g. when
+    // received from the backend), local modifications can bring elements out of order. We need to
+    // re-sort the elements to ensure that canonical IDs are independent of insertion order.
+    SortedMap<String, Value> sortedMap = new TreeMap<>(mapValue.getFieldsMap());
```

Comment:
Building a binary tree to sort the entries is pretty wasteful in terms of memory allocation. It's also pessimistic because building the tree is always going to cost O(n lg(n)). An alternative would be to collect the keys into an ArrayList and sort that. That has the benefit of allocating less memory and also runs in O(n) if the list is already sorted.

Reply:
I went through different implementations for optimize for both runtime complexity and readability. While I still think that using large Maps here is kind of an anti-pattern (especially given our lack of truncation support),  adding an extra line here to pre-sort the keys (and performing a key-based lookup on the HashMap) might be a worthy tradeoff. Updated.